Update README.md (#7)
lambdaclass/lambchain: README.md

# Bootcamp Verifying Lambchain (WIP)

An application-specific blockchain built using [Cosmos SDK](https://docs.cosmos.network/) and created with [Ignite CLI](https://ignite.com/). The blockchain offers a variety of zkSNARK implementations to verify proofs sent over transactions, and stores their results.

Cosmos SDK provides a framework to build an application layer on top of a consensus layer interacting via ABCI (Application BlockChain Interface). By default, [CometBFT](https://cometbft.com/) (a fork of Tendermint) is used in the consensus and network layer.

> Ignite CLI is used to generates boilerplate code for a Cosmos SDK application, making it easier to deploy a blockchain to production.

Reviewer: ```suggestion
Ignite CLI is used to generate boilerplate code for a Cosmos SDK application, making it easier to deploy a blockchain to production.
```

lambchaind tx lambchain verify --from alice --chain-id lambchain <proof>
```

This will output the transaction result (usually containing default values as it doesn't wait for the blockchain to execute it), and the transaction hash.

```txt
...
txhash: F105EAD99F96289914EF16CB164CE43A330AEDB93CAE2A1CFA5FAE013B5CC515
```

To get the transaction result, run:

```sh
lambchaind query tx <txhash>
```

## How It Works

### Project Anatomy

> The core of the state machine `App` is defined in [app.go](https://github.com/lambdaclass/lambchain/blob/main/app/app.go). The application inherits from Cosmos' `BaseApp`, which routes messages the the appropiate module for handling. A transaction contains any number of messages.

Reviewer: ```suggestion
The core of the state machine `App` is defined in [app.go](https://github.com/lambdaclass/lambchain/blob/main/app/app.go). The application inherits from Cosmos' `BaseApp`, which routes messages to the appropriate module for handling. A transaction contains any number of messages.
```

> The core of the state machine `App` is defined in [app.go](https://github.com/lambdaclass/lambchain/blob/main/app/app.go). The application inherits from Cosmos' `BaseApp`, which routes messages the the appropiate module for handling. A transaction contains any number of messages.

Author: already fixed

Cosmos SDK provides an Application Module interface to facilitate the composition of modules to form a functional unified application. Custom modules are defined in the [x](https://github.com/lambdaclass/lambchain/blob/main/x/) directory.

A module defines a message service for handling messages. These services are defined in a [protobuf file](https://github.com/lambdaclass/lambchain/blob/main/proto/lambchain/lambchain/tx.proto). The methods are then implemented in a [message server](https://github.com/lambdaclass/lambchain/blob/main/x/lambchain/keeper/msg_server.go), which is registered in the main application.

> Each message is identified by it's fully-qualified name. For example, the _verify_ message has the type `/lambchain.lambchain.MsgVerify`.

Reviewer: ```suggestion
Each message is identified by its fully-qualified name. For example, the _verify_ message has the type `/lambchain.lambchain.MsgVerify`.
```

A module usually defines a [keeper](https://github.com/lambdaclass/lambchain/blob/main/x/lambchain/keeper/keeper.go) which encapsulates the sub-state of each module, tipically through a key-value store. A reference to the keeper is stored in the message server to be accesed by the handlers.

<p align="center">

>   <img src="imgs/Diagram_Cosmos.svg">

Reviewer 2: @fborello-lambda Talking with @JulianGCalderon, we think that the _zk-SNARKs_ box should be called _lambchain_ since that is the name of our module.

</p>

The boilerplate for creating custom modules and messages can be generated using Ignite CLI. To generate a new module, run:

```sh
ignite scaffold module <module-name>
```

To generate a message handler for the module, run:

```sh
ignite scaffold message --module <module-name> <message-name> \
    <parameters...> \
    --response <response-fields...>
```

See the [Ignite CLI reference](https://docs.ignite.com/references/cli) to learn
about other scaffolding commands.

### Transaction Lifecycle

A transaction can be created and sent with protobuf with ignite CLI, using the following command:

```sh
lambchaind tx lambchain verify --from alice --chain-id lambchain "base64-encoded proof"
```

> A JSON representation of the transaction can be obtained with the `--generate-only` flag. It contains transaction metadata and a set of messages. A **message** contains the fully-qualified type to route it correctly, and it's parameters.

Reviewer: ```suggestion
A JSON representation of the transaction can be obtained with the `--generate-only` flag. It contains transaction metadata and a set of messages. A **message** contains the fully-qualified type to route it correctly, and its parameters.
```

- `checkTx`: The default `BaseApp` implementation does the following.

>     - Checks that a handler exists for every message based on it's type.

Reviewer: ```suggestion
    - Checks that a handler exists for every message based on its type.
```

>     - The `ValidateBasic` methods is executed for every message, allowing stateless validation.

Reviewer 2: This step is deprecated, I'd add a note about this
```suggestion
    - A `ValidateBasic` method (optionally implemented for each message type) is executed for every message, allowing stateless validation. This step is deprecated and should be avoided.
```

    - The `AnteHandler`'s are executed, by default verifying transaction authentication and gas fees.

> - `deliverTx`: In addition to the procedure previously mentioned.

Reviewer: It could be rephrased to make the meaning clearer.

> - `deliverTx`: In addition to the procedure previously mentioned.

Reviewer 2: ```suggestion
- `deliverTx`: In addition to the `checkTx` steps, previously mentioned, the following is executed to.
```
